### PR TITLE
executor: introduce the column buffer allocator for vectorized evaluation

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
In vectorized evaluation, some expressions need buffers to store their children's data.
This PR introduces an interface `columnBufferAllocator` to do this job.

Being compared with adding these buffers as expression's struct fields like `struct someBuiltinFunc{baseBuiltinFunc, buffer1, buffer2}`, this way has two advantages:
1. keep all expressions concurrency-safe,
2. it's convenient for us to optimize the strategy of memory management in the future. 

### What is changed and how it works?
1. Introduce a new interface `columnBufferAllocator`.
2. Implement this new interface with `localSliceBuffer`.
3. Add some unit tests and benchmarks.

Now, the `localSliceBuffer` is implemented by `Slice+Lock`, and the slight performance drawback seems acceptable:
```
BenchmarkPlusIntBufAllocator/enable-12           1000000              2089 ns/op               0 B/op          0 allocs/op
BenchmarkPlusIntBufAllocator/disable-12          1000000              1993 ns/op               0 B/op          0 allocs/op
```

**All changes are new code and unit tests, so it's easy to review.**

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/builtin_vectorized.go [hunk 1]
```diff
@@ -14,13 +14,97 @@
...
+	}
+	return r.newBuffer(evalType, capacity)
+}
+
+func (r *localSliceBuffer) release(buf *chunk.Column) {
+	r.Lock()
+	defer r.Unlock()
+	if r.size == len(r.buffers) {
+		buffers := make([]*chunk.Column, len(r.buffers)*2)
+		for pos, i := r.head, 0; i < len(r.buffers); i++ {
```
Comment (zz-jason) [inline expression/builtin_vectorized.go]:
```suggestion
		for pos, i := r.head, 0; i < len(r.buffers) && pos != r.tail; i++ {
```

File: expression/builtin_vectorized.go [hunk 2]
```diff
@@ -14,13 +14,97 @@
...
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/json"
 	"github.com/pingcap/tidb/util/chunk"
 )
 
+// columnBufferAllocator is used to allocate and release column buffer in vectorized evaluation.
+type columnBufferAllocator interface {
+	// alloc allocates a column buffer with the specific eval type and capacity.
+	// the allocator is not responsible for initializing the column, so please initialize it before using.
+	alloc(evalType types.EvalType, capacity int) (*chunk.Column, error)
```
Comment (zz-jason) [inline expression/builtin_vectorized.go]: can we rename these two fields to `put` and `get`?

File: expression/builtin_vectorized.go (hunk 1 above)
Comment (qw4990) [reply expression/builtin_vectorized.go]: If the pool is full, the tail and head are pointing to the same slot, so this check is not appropriate and only checking `i` is enough?

File: expression/builtin_vectorized.go (hunk 2 above)
Comment (qw4990) [reply expression/builtin_vectorized.go]: These two methods are renamed.

File: expression/builtin_vectorized.go [hunk 3]
```diff
@@ -14,13 +14,97 @@
...
+	r.Lock()
+	defer r.Unlock()
+	if r.size > 0 {
+		buf := r.buffers[r.head]
+		r.head++
+		if r.head == len(r.buffers) {
+			r.head = 0
+		}
+		r.size--
+		return buf, nil
```
Comment (zz-jason) [inline expression/builtin_vectorized.go]: Should we call `buf.Reserve()` before returning it?

File: expression/builtin_vectorized.go (hunk 1 above)
Comment (zz-jason) [reply expression/builtin_vectorized.go]: okay, how about `i < r.size`?

Comment (qw4990) [reply expression/builtin_vectorized.go]: That's ok.

File: expression/builtin_vectorized.go (hunk 3 above)
Comment (qw4990) [reply expression/builtin_vectorized.go]: There are multiple ways to initialize a buffer (`Resize`/`Reset`/`Reserve`) and in this method, the allocator doesn't know which way is better.
Therefore it seems more appropriate to let users initialize these buffers themselves.

File: expression/builtin_vectorized.go (hunk 1 above)
Comment (qw4990) [reply expression/builtin_vectorized.go]: This loop has been optimized to two `slice-copy` operations.

File: expression/builtin_vectorized.go [hunk 4]
```diff
@@ -14,13 +14,92 @@
...
 	"github.com/pingcap/errors"
+	"github.com/pingcap/parser/mysql"
 	"github.com/pingcap/tidb/sessionctx"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/json"
 	"github.com/pingcap/tidb/util/chunk"
 )
 
+// columnBufferAllocator is used to allocate and release column buffer in vectorized evaluation.
+type columnBufferAllocator interface {
```
Comment (XuHuaiyu) [inline expression/builtin_vectorized.go]: Will it be better to define this in util package?

File: expression/builtin_vectorized.go [hunk 5]
```diff
@@ -14,13 +14,92 @@
...
+	sync.Mutex
+	buffers []*chunk.Column
+	head    int
+	tail    int
+	size    int
+}
+
+func newLocalSliceBuffer() *localSliceBuffer {
+	// in most cases, expressions are called serially, so 1 buffer is enough.
+	return &localSliceBuffer{buffers: make([]*chunk.Column, 1)}
```
Comment (XuHuaiyu) [inline expression/builtin_vectorized.go]: 1 seems to be not enough if a function has multiple args.

File: expression/builtin_vectorized.go [hunk 6]
```diff
@@ -14,13 +14,92 @@
...
+// columnBufferAllocator is used to allocate and release column buffer in vectorized evaluation.
+type columnBufferAllocator interface {
+	// get allocates a column buffer with the specific eval type and capacity.
+	// the allocator is not responsible for initializing the column, so please initialize it before using.
+	get(evalType types.EvalType, capacity int) (*chunk.Column, error)
+	// put releases a column buffer.
+	put(buf *chunk.Column)
+}
+
+type localSliceBuffer struct {
```
Comment (XuHuaiyu) [inline expression/builtin_vectorized.go]: Add some comment for it.

File: expression/builtin_vectorized.go (hunk 4 above)
Comment (qw4990) [reply expression/builtin_vectorized.go]: It's only used in vectorized evaluation now so keeping it here is acceptable now.

File: expression/builtin_vectorized.go (hunk 5 above)
Comment (qw4990) [reply expression/builtin_vectorized.go]: You are right.
Now I add a new argument `initCap` for this function, which is decided by the number of arguments.

File: expression/builtin_vectorized.go (hunk 6 above)
Comment (qw4990) [reply expression/builtin_vectorized.go]: Some comments are added.

**Comment by qw4990** [comment]
All comments are addressed, PTAL @zz-jason @XuHuaiyu

**Pull request merged**